app2/ota_metadata.py

```diff
@@ -129,70 +165,47 @@ def _parse_metadata(self, metadata_jwt):
         return jwt_list[0] + "." + jwt_list[1]
 
-    def get_directories_info(self):
-        """
-        return
-            directory file info list: { "file": file name, "hash": file hash }
-        """
-        return self.__metadata_dict["directory"]
-
-    def get_symboliclinks_info(self):
-        """
-        return
-            symboliclink file info: { "file": file name, "hash": file hash }
-        """
-        return self.__metadata_dict["symboliclink"]
-
-    def get_regulars_info(self):
-        """
-        return
-            regular file info: { "file": file name, "hash": file hash }
-        """
-        return self.__metadata_dict["regular"]
-
-    def get_persistent_info(self):
-        """
-        return
-            persistent file info: { "file": file name, "hash": file hash }
-        """
-        return self.__metadata_dict["persistent"]
-
-    def get_rootfsdir_info(self):
-        """
-        return
-            rootfs_directory info: {"file": dir name }
-        """
-        return self.__metadata_dict["rootfs_directory"]
-
-    def get_certificate_info(self):
-        """
-        return
-            certificate file info: { "file": file name, "hash": file hash }
-        """
-        return self.__metadata_dict["certificate"]
+    def _verify_certificate(self, certificate: str):
+        certs = [cert.name for cert in sorted(list(self._certs_dir.glob("*.*.pem")))]
+        if len(certs) == 0:
+            logger.warning("there is no root or intermediate certificate")
+            return
+        # e.g. certs == [A.1.pem, A.2.pem, B.1.pem, B.2.pem]
+        prefixes = set()
+        for cert in certs:
+            m = re.match(r"(.*)\..*.pem", cert)
+            prefixes.add(m.group(1))
+        logger.info(f"certs prefixes {prefixes}")
+
+        def _load_certificates(store, certs):
+            for cert in certs:
+                logger.info(f"cert {cert}")
+                c = crypto.load_certificate(crypto.FILETYPE_PEM, open(cert).read())
+                store.add_cert(c)
+
+        for prefix in sorted(prefixes):  # use sorted to fix the order of verification
+            store = crypto.X509Store()
+            _load_certificates(store, self._certs_dir.glob(f"{prefix}.*.pem"))
+            try:
+                cert_to_verify = crypto.load_certificate(
+                    crypto.FILETYPE_PEM, certificate
+                )
+            except crypto.Error as e:
+                raise OtaErrorRecoverable(f"invalid certificate {certificate}")
+
+            try:
+                store_ctx = crypto.X509StoreContext(store, cert_to_verify)
+                store_ctx.verify_certificate()
+                logger.info("certificate verify: OK")
+                return
+            except crypto.X509StoreContextError as e:
+                logger.warning(f"NOTE: This warning can be ignored {e}")
+
+        raise OtaErrorRecoverable(f"certificate {certificate} could not be verified")
```

Comment:
```suggestion
    def _verify_certificate(self, certificate: str):
        ca_set_prefix = set()
        for cert in self._certs_dir.glob(f"*.*.pem"):
            m = re.match(r"(.*)\..*.pem", cert.name)
            ca_set_prefix.add(m.group(1))

        for ca_prefix in ca_set_prefix:
            certs_list = [self._certs_dir/c.name for c in self._certs_dir.glob(f"{ca_prefix}.*.pem")]
            load_pem = partial(crypto.load_certificate, crypto.FILETYPE_PEM)

            store = crypto.X509Store()
            for c in certs_list:
                store.add_cert(load_pem(open(c).read()))

            cert_loaded = load_pem(open(certificate).read())
            try:
                store_ctx = crypto.X509StoreContext(store, cert_loaded)
                store_ctx.verify_certificate()
                logger.info(f"verfication succeeded against: {ca_prefix}")
                return
            except crypto.X509StoreContextError as e:
                logger.debug(f"verify against {ca_prefix} failed: {e}")

        raise OtaErrorRecoverable(f"certificate {certificate} could not be verified")
```

Reply:
A small improvement over the previous implementation. 
We don't need to sort the certs_list as the crypto.X509Store will handle it for us.
Exception handling is also improved.